### ISSUE TITLE
irods resource plugin using libcephfs2

### ISSUE THREAD

**Issue by holtgrewe** [irods/irods#5919]
## Feature

CephFS has become really nice recently and nowadays offers interesting features including active-active mds servers etc.

iRODS has support for CephFS through the unix file system resource when the irods host has the ceph file system mounted. The user has the choice between using kernel mounts and FUSE Ceph mounts. Kernel mounts offer better performance while FUSE mounts can be considered more robust in the case that a crash of the ceph driver does not take down the Kernel.

A good middle ground would be to use libcephfs [as nfs-ganesha does](https://github.com/nfs-ganesha/nfs-ganesha/tree/next/src/FSAL/FSAL_CEPH). This way, nfs-ganesha can serve data from a ceph file system without mounting on the cluster.

Probably the greatest advantage of using libcephfs would be that iRODS could serve from CephFS file system without mounting when run inside a Docker container. In the case of problems with the Ceph system, the host would be unaffected and "only" the iRODS process would suffer.

Using librados as the backend is something alltogether.

**Comment by trel** [comment]
So you're proposing a new resource plugin, perhaps named... `cephfs` that would use libcephfs and have a better performance profile than using the existing `unixfilesystem`.   Is that correct?

**Comment by holtgrewe** [comment]
Yes, that is correct. It is my understanding that the performance will be better than having CephFS mounted via FUSE and not risk any kernel instability as you have with CephFS kernel mounts. Also, having libceph in the user space gives you the advantage that you can update the CephFS client version by rebuilding the resource module rather than upgrading with your kernel.

Some background:

- The Ceph *server* implements the data sharing in user space. There is a new major release every year and the continually add features (such as more advanced rebalancing).
- The Ceph *client* can mount via the built-in kernel module (similar to mounting NFS). If the network or ceph server fails then your client kernel will not halt (as I have experienced with NFSv3) but commands like `df` will stall. An upgrade of the client can then only be done by upgrading your kernel. If you run old kernels that the server cannot enable certain features as the ceph client has not been updated yet. This problem is mitigated by Red Hat back-porting the ceph driver into their kernels.
- A second option is the FUSE module. This is written in Python and people report something like 1/10th of IOPs performance of the kernel module.

To summarize, it combines the best of both Ceph FUSE and kernel mounts and would simplify container-based deployment a bit.

**Comment by kript** [comment]
As a counter argument,  rados is much more mature as an access protocol and server implementation then cephfs; so using the [iRods rados plugin](https://github.com/irods/irods_resource_plugin_rados) feels like a far safer option then using cephfs.
